### ISSUE TITLE
"Cannot look up subprotocol when connection is closing"

### ISSUE THREAD

**Issue by lithp** [ethereum/trinity#1367]
* Trinity version: 190ab8a7d23b8b02e4998e89fdcf48cb1ac776ca

### What is wrong?

Trinity syncs, but crashes after about an hour of activity with this exception which causes a cancellation to propagate up through everything:

```
ESC[1mESC[33m WARNING  2019-12-07 15:31:04,328           ETHPeerPool  Task <coroutine object BasePeerPool._periodically_report_stats at 0x7f0f10b99648> finished unexpectedly: PeerConnectionLost('Cannot look up s
ubprotocol when connection is closing')ESC[0m
   DEBUG  2019-12-07 15:31:04,328           ETHPeerPool  Task failure traceback
Traceback (most recent call last):
  File "/home/brian/trinity/p2p/service.py", line 158, in _run_task_wrapper
    await awaitable
  File "/home/brian/trinity/p2p/service.py", line 179, in _run_daemon_task_wrapper
    await awaitable
  File "/home/brian/trinity/p2p/peer_pool.py", line 537, in _periodically_report_stats
    if not hasattr(peer, "eth_api"):
  File "/home/brian/.local/share/virtualenvs/trinity-dwQEboUK/lib/python3.7/site-packages/cached_property.py", line 35, in __get__                                                                                
    value = obj.__dict__[self.func.__name__] = self.func(obj)
  File "/home/brian/trinity/trinity/protocol/eth/peer.py", line 86, in eth_api
    return self.connection.get_logic(ETHAPI.name, ETHAPI)
  File "/home/brian/trinity/p2p/connection.py", line 228, in get_logic
    if not self.has_logic(name):
  File "/home/brian/trinity/p2p/connection.py", line 224, in has_logic
    raise PeerConnectionLost("Cannot look up subprotocol when connection is closing")
p2p.exceptions.PeerConnectionLost: Cannot look up subprotocol when connection is closing
```

I can put up a more detailed log if that would be helpful.

**Comment by lithp** [comment]
After running across this I tried using `v0.1.0-alpha.31` and haven't run into any issues, so it might have been introduced by the 22 commits which have been committed since then?

```bash
brian@pop-os:~/trinity$ git log --oneline v0.1.0-alpha.31..190ab8a7d23b8b02e4998e89fdcf48cb1ac776ca
190ab8a7 (upstream/master, master) Merge pull request #1355 from carver/quiet-broken-pipe-errors
ed6663d5 Merge pull request #1365 from cburgdorf/christoph/feat/matomo-tracking
0471b6e1 Add matomo tracking to docs site
6a2c5be7 Merge pull request #1364 from carver/mutual-peer-connection
a82c4325 Merge pull request #1354 from carver/squash-already-cancelled-logs
dd440fae Merge pull request #1356 from carver/cleaner-txpool-shutdown
aba4d023 Suppress BrokenPipeError from previews on shutdown
d454f9c0 Add exception suppressor, and bugfix the async one
05ae1758 Show PeerConnection instead of UnknownAPI on close
2b20b4b1 Avoid double-connection when peers mutually dial
80e96f2b Add repr and str to p2p.Connection
af451b51 Merge pull request #1360 from ralexstokes/fork-choice-updates-for-v0.9.2-cherry-pick
39a9e49d PR feedback for #1357
ee27212d Add stack trace to logs, when service cancelled 2x
9ff9e63e Silence warnings of connection "already cancelled"
be619f12 Add release notes #1356
12a6a739 show warning when skipping ahead for beam sync
29b9aebc Cleaner TxPool shutdown
53b98510 Add fork choice spec updates from v0.9.2
08892c96 Convert DiscoveryService to use the trio-based Service
121e2ccf Merge pull request #1357 from carver/just-linter-fixups
602b91cf Mysterious unrelated linter fixups
brian@pop-os:~/trinity$ git log --oneline v0.1.0-alpha.31..190ab8a7d23b8b02e4998e89fdcf48cb1ac776ca | wc -l
22
```

**Comment by carver** [comment]
If it wasn't fixed before, it was probably fixed by one of #1545 or #1548 . I'll optimistically close, until it comes back again.
